ping/rust/src/bin/testplan_0500.rs

```diff
@@ -37,18 +51,23 @@ struct OrphanRuleWorkaround(Swarm<Behaviour>);
 
 #[async_trait]
 impl PingSwarm for OrphanRuleWorkaround {
-    async fn listen_on(&mut self, address: &str) -> Result<()> {
+    async fn listen_on(&mut self, address: &str) -> Result<Option<String>> {
```

Comment:
The Option is because none of the existing tests return an address and I didn't want to update them all (and the signature must be the same across versions for the trait).

In older tests the address being listened on was also dialable. The return value is really only helpful for WebRTC with the `/certhash/` business.

Reply:
In theory, the returned address is also useful if you listened on port 0 for other transports as this address will contain the actual port chosen by the kernel.

If you look at the other tests, you end up defaulting to the passed in address so it will be cleaner if we just always return the address here because there is one less conditional operation.

Reply:
It's not conditional in the old tests (everything prior to testplan_0500). Their versions of this function always returns None in the Ok case, and test always ignores it.

I could make the newer tests always rely on this function and bail if they got None. So, one less conditional there.

Or, I can change the trait to return `Result<String>` and convert all the old implementations of listen_on to return an address. And then, if you like, we could change the old tests to use this, or continue ignoring it.

Since they already work, and the goal was to test new functionality, I was hesitant to change the old tests more than necessary. But whatever works.

Reply:
> It's not conditional in the old tests

Matching on an `Option` sounds pretty conditional to me :man_shrugging: 

> Or, I can change the trait to return `Result<String>` and convert all the old implementations of listen_on to return an address. And then, if you like, we could change the old tests to use this

That would be my preference, yes.

> Since they already work, and the goal was to test new functionality, I was hesitant to change the old tests more than necessary.

That is a good goal overall. I created the `PingSwarm` abstraction with the back-then usecases in mind. It is however not set in stone and I'd prefer to iterate on it to keep as much functionality as possible shared between the tests.

It is not ideal that we have to maintain old tests. My personal long-term goal is to do something like https://github.com/testground/testground/issues/1522 to fix this where "old tests" are somewhere in our Git history and it is only the back-then created docker containers that are still around and used for the tests.